src/main/java/edu/kit/compiler/codegen/PatternCollection.java

```diff
@@ -0,0 +1,159 @@
+
+                Map.entry(iro_Div, new DivisionPattern(Type.DIV, REGISTER, REGISTER)),
+                Map.entry(iro_Mod, new DivisionPattern(Type.MOD, REGISTER, REGISTER)),
+
+                Map.entry(iro_Minus, new UnaryInstructionPattern(iro_Minus, "neg", REGISTER, true, false)),
+
+                Map.entry(iro_Cmp, new ComparisonPattern()),
+
+                Map.entry(iro_Conv, new ConversionPattern()),
+
+                Map.entry(iro_Store, new BinaryInstructionPattern(iro_Store, "mov", MEMORY, REGISTER, false, true)),
+                Map.entry(iro_Load, new LoadMemoryPattern()),
+
+                Map.entry(iro_Call, new CallPattern()),
+                Map.entry(iro_Phi, new PhiPattern()),
+                Map.entry(iro_Block, new BlockPattern()),
+                Map.entry(iro_Return, new ReturnPattern()),
+
+                Map.entry(iro_Jmp, new ConditionPattern()),
+                Map.entry(iro_Cond, new ConditionPattern()),
+
+                // trivial pattern for nodes without predecessors
+                Map.entry(iro_Start, new EmptyPattern()),
+                Map.entry(iro_End, new EmptyPattern()),
+                Map.entry(iro_Address, new EmptyPattern()));
+    }
+
+    private static final class EmptyPattern implements Pattern<InstructionMatch> {
+        @Override
+        public InstructionMatch match(Node node, MatcherState matcher) {
+            return InstructionMatch.empty(node, StreamSupport
+                    .stream(node.getPreds().spliterator(), false)
+                    .collect(Collectors.toList()));
+        }
+    }
+
+    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+    private static final class CompoundPattern implements Pattern<InstructionMatch> {
+
+        private final List<Pattern<InstructionMatch>> patterns;
+
+        @Override
+        public InstructionMatch match(Node node, MatcherState matcher) {
+            for (var pattern : patterns) {
+                var shadow = new MatcherShadow(matcher);
+                var match = pattern.match(node, shadow);
+                if (match.matches()) {
+                    shadow.merge();
+                    return match;
+                }
+            }
+            return InstructionMatch.none();
+        }
+    }
+
+    private static final class MatcherShadow extends MatcherState {
+
+        private final MatcherState subject;
+
+        public MatcherShadow(MatcherState subject) {
+            super(subject.graph, subject.registerCount);
+            this.subject = subject;
+        }
+
+        public void merge() {
+            subject.phiRegisters.putAll(this.phiRegisters);
+            subject.registerCount = this.registerCount;
+        }
+
+        @Override
+        public InstructionMatch getMatch(Node node) {
+            return subject.getMatch(node);
+        }
+
+        @Override
+        public int getNewRegister() {
+            return super.getNewRegister();
```

Comment:
Isnt't this equivalent to not overriding the method at all?

Reply:
Yes, this is already fixed locally.